pytorch_lightning/core/decorators.py

```diff
@@ -62,3 +63,42 @@ def auto_transfer_args(self, *args, **kwargs):
         return fn(self, *args, **kwargs)
 
     return auto_transfer_args
+
+
+def parameter_validation(fn: Callable) -> Callable:
+    """
+    Decorator for :meth:`~pytorch_lightning.core.LightningModule.to` method.
+    Validates that the module parameter lengths match after moving to the device. It is useful
+    when tying weights on TPU's.
+
+    Args:
+        fn: ``.to`` method
+
+    Note:
+        TPU's require weights to be tied/shared after moving the module to the device.
+        Failure to do this results in the initialization of new weights which are not tied.
+        To overcome this issue, weights should be tied using the ``on_post_move_to_device`` model hook
+        which is called after the module has been moved to the device.
+
+    See Also:
+        - `XLA Documentation <https://github.com/pytorch/xla/blob/master/TROUBLESHOOTING.md#xla-tensor-quirks>`_
+    """
+
+    @wraps(fn)
+    def inner_fn(self, *args, **kwargs):
+        pre_layer_count = len(list(self.parameters()))
+        module = fn(self, *args, **kwargs)
+        self.on_post_move_to_device()
+        post_layer_count = len(list(self.parameters()))
+
+        if not pre_layer_count == post_layer_count:
+            rank_zero_warn(
```

Comment:
@lezwon Could you help me out on what this check means?

Reply:
@kaushikb11 if the layer weights are not tied while on tpu, then the layer count on the tpu will show as  no_of_layers + 1, as the xla library will make a copy of the layer weights. This check makes sure the layer count matches after moving the model to the device. 😊👍

Reply:
@lezwon Yup, but I don't think it's necessary for doing `parameter_validation` for every module `to` call? wdyt?

Just a fyi: Currently I am doing this https://github.com/PyTorchLightning/pytorch-lightning/blob/tpu_spawn_added/pytorch_lightning/plugins/training_type/tpu_spawn.py#L172, by using xla's MpModelWrapper. This will make the test fail. Maybe I could add this check specific to TPU acclerators.

Reply:
Yes, you could maybe add it specifically to TPU accelerators. Also do check it works well when using xla with GPU's.